### PR TITLE
Automated flow for GDPR/CCPA deletion requests

### PR DESCRIPTION
**Pull request by cmyui**
Tested locally, generally seems to be working as intended.

### CODE REVIEW

File: app/api/internal/v1/users.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    ErrorCode.INSUFFICIENT_PRIVILEGES: 401,
+    ErrorCode.PENDING_VERIFICATION: 401,
+    ErrorCode.NOT_FOUND: 404,
+    ErrorCode.CONFLICT: 409,
+    ErrorCode.INTERNAL_SERVER_ERROR: 500,
+}
+
+
+@router.delete("/api/v1/users/{user_id}")
+async def delete_user(user_id: int) -> Response:
```
Comment (cmyui) [inline app/api/internal/v1/users.py:33]: To be invoked from RAP

File: app/usecases/users.py [hunk 2]
```diff
@@ -192,3 +200,142 @@ async def update_email_address(
 
 async def fetch_total_registered_user_count() -> int:
     return await users.fetch_total_registered_user_count()
+
+
+async def delete_one_by_user_id(user_id: int, /) -> None | Error:
+    """\
+    An anonymization process for user deletion, mainly implemented
+    for the purpose of complying with GDPR, CCPA and other regulations.
+    """
```
Comment (cmyui) [inline app/usecases/users.py:209]: There are still quite a few open questions for "full compliance", but this is a pretty good start

File: app/usecases/users.py [hunk 3]
```diff
@@ -192,3 +200,142 @@ async def update_email_address(
...
+                        # no other members in the clan; just delete it
+                        await clans.delete_one_by_clan_id(user.clan_id)
+
+        await password_recovery.delete_many_by_username(user.username)
+
+        # TODO: consider what ac data should be anonymized instead of wiped
+        await user_ip_associations.delete_many_by_user_id(user_id)
+        await user_hwid_associations.delete_many_by_user_id(user_id)
+        await lastfm_flags.delete_many_by_user_id(user_id)
+        # TODO: patcher_detections & patcher_token_logs
```
Comment (cmyui) [inline app/usecases/users.py:305]: @tsunyoku do you know how can i identify records from a given user in these?

File: app/usecases/users.py [hunk 4]
```diff
@@ -192,3 +200,142 @@ async def update_email_address(
...
+        )
+        await transaction.rollback()
+        return Error(
+            error_code=ErrorCode.INTERNAL_SERVER_ERROR,
+            user_feedback="Failed to process user deletion request.",
+        )
+    else:
+        logging.info(
+            "Successfully processed GDPR/CCPA user deletion request",
+            extra={"user_id": user_id},
```
Comment (cmyui) [inline app/usecases/users.py:338]:
```suggestion
            # NOTE: intentionally not logging any pii
            extra={"user_id": user_id},
```

File: app/api/internal/v1/users.py (hunk 1 above)
Comment (cmyui) [reply app/api/internal/v1/users.py:33]: Probably with a ratelimit on admins, e.g. 3/hr or 10/d or something

File: app/usecases/users.py [hunk 5]
```diff
@@ -192,3 +200,143 @@ async def update_email_address(
...
+    # - [leave as-is] user_achievements
+    # - [transfer perms if owner & kick] clans
+    # - [leave as-is] identity_tokens
+    # - [leave as-is] irc_tokens
+    # - [TODO/AC] lastfm_flags
+    # - [leave as-is] beatmaps_rating
+    # - [leave as-is] clan_requests (empty?)
+    # - [leave as-is] comments
+    # - [leave as-is] matches
+    # - [leave as-is] match_eventrs
```
Comment (cmyui) [inline app/usecases/users.py]:
```suggestion
    # - [leave as-is] match_events
```

File: app/usecases/users.py (hunk 3 above)
Comment (tsunyoku) [reply app/usecases/users.py:305]: both tables have a `score_id` column you can do a join by. you'll also want to clear `score_submission_logs`, which also has a `score_id` column

**Comment by infernalfire72** [comment]
dont clear hardware ids please (just set the associated user to 0 maybe or anonymize the actual user)

**Comment by cmyui** [comment]
> dont clear hardware ids please (just set the associated user to 0 maybe or anonymize the actual user)

I want to get this feature our ASAP & there are some other code changes required here to accomodate this desire, so I think we can treat this as p3/follow-up, so long as we're only really deleting unbanned users for the time being. We _do_ have the follow note in our account deletion acceptance terms:
```
4. If you are currently banned from our services, your device ban will be unassociated from the user account, and your ban will become un-appealable.
```

File: app/usecases/users.py (hunk 3 above)
Comment (cmyui) [reply app/usecases/users.py:305]: I see `patcher_token_logs` has a score id, ty, and noted on the `score_submission_logs`, will do - but `patcher_detections` doesn't seem to include any ids I recognize. Do you know what I could use?

Comment (cmyui) [reply app/usecases/users.py:305]: Going to merge for now and revisit this

**Pull request merged**